File: Depth-First Search/Binary tree/Find paths/LET_no968_Binary Tree Cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int minCameraCover(TreeNode* root) {
        return (dfs(root) == "toBeCovered") + res;
    }
    string dfs(TreeNode* node) {
        if (!node)
            return "covered";
        
        string left = dfs(node->left), right = dfs(node->right);
        if (left == "toBeCovered" || right == "toBeCovered") {
            res++;
            return "covering";
        }
        return (left == "covering" || right == "covering" ? "covered" : "toBeCovered");
    }
};

File: Depth-First Search/Binary tree/Find paths/LET_no968_Binary Tree Cameras.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minCameraCover(self, root: TreeNode) -> int:
        self.res = 0
        
        def dfs(node):
            if not node: return "coverd"

            left, right = dfs(node.left), dfs(node.right)
            if left == "toBeCovered" or right == "toBeCovered":
                self.res += 1
                return "covering"
            return "covered" if left == "covering" or right == "covering" else "toBeCovered"
    
        return (dfs(root) == "toBeCovered") + self.res
    
    